rust/foxglove/src/app_url.rs

```diff
@@ -0,0 +1,114 @@
+//! App URL builder
+
+use std::borrow::Cow;
+use std::fmt::Display;
+
+static BASE_URL: &str = "https://app.foxglove.dev";
+
+#[derive(Debug, Clone)]
+enum DataSource {
+    WebSocket(String, u16),
+}
+
+/// Foxglove app URL.
+#[must_use]
+#[derive(Debug, Default, Clone)]
+pub struct AppUrl {
+    data_source: Option<DataSource>,
+    layout_id: Option<String>,
+    open_in_desktop: bool,
+}
+
+impl Display for AppUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(BASE_URL)?;
+        for (ii, (k, v)) in self.query_params().into_iter().enumerate() {
```

Comment:
There is also the url crate we could have used for this, but I think given our needs at present are very minimal, this is better, as it doesn't bring additional dependencies.